### DISCUSSION TITLE
Why add strange syntax sugar to C#?

### DISCUSSION THREAD

**Discussion by reitowo** [dotnet/csharplang#5287] [Language Ideas]
There're several new language feature start from C# 9 which is strange, and breaks C#'s these years strict style. 
In fact, there're several good things such as default interface implementation, pattern matching, using declaration, etc. 
But the following introduce strange syntax sugar and creates 'both ways are OK' problems, which may cause misleading the beginners that may confused about 'which way is better', and also decreased the integrity of grammar:
1. Top-level statements
2. Global using directives
3. Record struct/class
4. File-scoped namespace declaration

**Comment by HaloFour** [comment]
It's up to the language design team to deem whether such features are worth adding to the language, and in this case they did.  Those features have shipped or are very close to shipping, and they won't be removed.

It's inevitable that some people will disapprove of language changes, and they are welcome to their opinion, but language design does not require unanimous support.  If it did no language anywhere would ever be designed, let alone evolve.

**Reply by reitowo** [reply]
Just feeling C# is more like a community based language now, It was more likely an Microsoft ruled design. Certain quantity of people's need is enough for the design team to take into account in next version's design, while C++ refuses to change even some feature misleads most of people.

**Reply by HaloFour** [reply]
Every feature you mentioned was requested by the community.  Design discussions were carried out on this repository and in the public, with community feedback considered.

The community has a very wide array of opinions as to how the language should evolve.  At the end of the day it is the Language Design Team at Microsoft that are the final arbiters.

**Reply by reitowo** [reply]
Yes, I think starting from C# 8 the design was moved from official to community.

**Reply by HaloFour** [reply]
> Yes, I think starting from C# 8 the design was moved from official to community.

C# moved to an open source repo at version 6, but the decisions as to what features are added and how has always been decided by the language team.  Community feedback is important, but the language design process is not a democracy.

**Comment by CyrusNajmabadi** [comment]
> the beginners that may confused about 'which way is better',

Beginners can be taught. I'm not seeing why that is a problem. Indeed, the conclusion of this approach to design would be that we could never improve things as that will give beginners something extra to learn and that will question which way is better.

**Reply by MgSam** [reply]
I think it's a pretty undeniable fact that the more ways a language offers to do the same things that there is some unavoidable level of confusion that its going to cause. C++, of course, being the textbook example. 

I've still seen developers in 2021, both young and old, writing new code that uses anonymous delegate syntax rather than lambdas. I routinely see developers, including at Microsoft, write out type names rather than using `var`.

**Reply by CyrusNajmabadi** [reply]
>  I routinely see developers, including at Microsoft, write out type names rather than using var.

The introduction of var was never to get all developers to stop writing out type names. Var was introduced because there were literally types without names and there was no way to write a variable type for them. 

Since the introduction of var we've been very intentional and clear that, outside of where the language requires it, it is completely optional and entirely a matter of personal (or team) taste.

**Reply by foxesknow** [reply]
> Beginners can be taught. 

Yes, and yet still top level statements were added!

**Reply by DavidArno** [reply]
@foxesknow,

I've been using C# for nearly 20 years and I'm loving those top level statements. All that nonsense of a static class and method just to get to the start of my code has gone, never to return. It's wonderful.

I guess it'll be useful for newbies too. 😆

**Comment by MgSam** [comment]
While I agree that certain features added in the last few versions feel like new syntax for new syntax's sake, I don't really agree with your list of examples. All the features you mention are serving the same goal of trying to reduce some of the noise and boilerplate to write C# code.

**Reply by jnm2** [reply]
I'm pretty sure there's not a person at the LDM who likes the thought of new syntax for the sake of new syntax, so the feeling is localized.

**Comment by Rekkonnect** [comment]
These new language features that essentially provide an alternative way to do the same thing, either fully replace the old way that bore some minor flaw, or specifically improve the majority of its use case into a simpler structure.

In that case, we have a language that evolves over time without risking flexibility. I don't see any problem with that other than more conservative users getting mind-boggled, which is a problem of their own, as they can still opt-out of the new features, and keep using the old language versions.

**Reply by jmarolf** [reply]
> These new language features that essentially provide an alternative way to do the same thing,

the language was Turing complete in C# 1.0 so this statement is true of every language feature ever added to C#. Its all just alternative syntax to do what you could already do in C# 1.0 :)

**Comment by ziaulhasanhamim** [comment]
I certainly disagree with you on this point. These are not strange. This makes c# code easier to read and write. For example, the file scoped namespaces. Just try to remember what is the time when you have seen two namespace declarations in one file. I don't think you have. So if there is only one namespace per file so it's very logical to make it file scoped instead of creating an extra code block for that. The same goes for top-level statements. I see nothing strange about these. And people keep mentioning that It would be tough for beginners. Yeah, sometimes it's true. But oftentimes the new way with new features is easy than the old ones(for example top-level statements). At the end of the day, it's improving the development experience. Also, what makes C# different from other languages is its capability to fit into any programming situation pretty nicely.

**Comment by softlion** [comment]
I agree with you, but not on your examples.

Top-level statements => very good for starters and scripts.  
Global using directives => very good for starters, they don't need to search in which lib a type is in anymore. Leaving more brain for something else. Resharper does that automagically since a couple of years now  
Record struct/class => i agree with you. But it's for perf and react like state usage. Like `Memory<T>` and `Span<T>`  
File-scoped namespace declaration => very good  for starters. Less space on the left margin. Useless anyway as the correct way is to have only one namespace per file.

**Reply by CyrusNajmabadi** [reply]
> I agree with you, but not on your examples.

And that's the rub in a nutshell.  Every user (and designer) is going to have their own personal set of things that they consider important and some they consider irrelevant.  This is a truism in the language as we've never designed a *single* feature that avoided critique from some group as being irrelevant or inappropriate to add.  

We could either do nothing in the language to make things more convenient (which would aggravate another set of users as well), or we can accept that this is a subjective area and that no matter what we do we will always have some group that is displeased.  

As you can see from just the people participating here, some of these features are absolutely loved (and our own feedback channels show this extends out broadly).  So, IMO, this was a solid thing for us to do.  Features that practically everyone in the community can easily adopt that just continue helping to lower friction and make things more enjoyable. 

I think it's safe to say that the LDM considers these actions to have been worthwhile, and we're going to continue doing them, even though we know there will be a small contingent that is always unhappy about any particular thing we do.

**Reply by softlion** [reply]
So i can still agree without receiving flames ? Ty !

**Reply by jnm2** [reply]
@softlion So far it looks like you can. If someone comes along and acts unprofessionally, it will be addressed.

**Reply by 333fred** [reply]
We want csharplang to be a place for _discussion_. While discussion can certainly be had between people who completely agree with each other, that's usually not a super meaningful discussion :). We much prefer discussions that have many viewpoints, as they bring out new ideas and perspectives that we might not have seen or considered before.
